File: 3/js/src/Quickselect.ts
function quickselect(list: number[], k: number): number {
	let first = 1;
	let last = list.length - 1;

	while (true) {
		let pivot = first - 1;

		while (first <= last) {
			if (list[first] > list[pivot] && list[last] < list[pivot]) {
				swap(first, last, list);
			}

			if (list[first] <= list[pivot]) {
				first++;
			}
			if (list[last] >= list[pivot]) {
				last--;
			}
		}

		swap(last, pivot, list);

		if (k - 1 == last) {
			return list[k - 1];
		}

		if (last < k - 1) {
			first = last + 2;
			last = list.length - 1;
		} else {
			first = 1;
			last--;
		}
	}
}

function swap(i: number, j: number, list: number[]) {
	[list[i], list[j]] = [list[j], list[i]];
}

(function main() {
	const list = [8, 5, 2, 9, 7, 6, 3];
	const k = 3;

	console.log(quickselect(list, k));
})();

File: 3/java/Quickselect.java
class Quickselect {
	public static void swap(int i, int j, int arr[]) {
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}

	public static int quickSelect(int arr[], int k) {
		int left = 1;
		int right = arr.length-1;
		
		while(true) {
			int pivotIdx = left-1;		
			
			while(left <= right) {
				if(arr[left] > arr[pivotIdx] && arr[right] < arr[pivotIdx]) {
					swap(left, right, arr);
				} 

				if(arr[left] <= arr[pivotIdx]) {
					left++;
				} 

				if(arr[right] >= arr[pivotIdx]) {
					right--;
				}
			}

			swap(pivotIdx, right, arr);

			if(right == k-1) {
				return arr[right];
			} else if(right < k-1) {
					left = right+2; 	
					right = arr.length-1;
			} else {
					left = 1;
					right -= 1;
			}
		}
	}

	public static void main(String args[]) {
		int arr[] = {8,5,2,9,7,6,3};
		int k = 3;
		
		System.out.println(quickSelect(arr, k));
	}
}
